internal/storage/index_data_codec.go

```diff
@@ -0,0 +1,375 @@
+	indexParamBlob, err := codec.SerializeIndexParams(indexBuildID, version, collectionID, partitionID, segmentID, fieldID, indexParams, indexName, indexID)
+	if err != nil {
+		return nil, err
+	}
+	blobs = append(blobs, indexParamBlob)
+
+	for pos := range datas {
+		blob, err := codec.serializeImpl(indexBuildID, version, collectionID, partitionID, segmentID, fieldID, indexName, indexID, datas[pos].Key, datas[pos].Value, ts)
+		if err != nil {
+			return nil, err
+		}
+		blobs = append(blobs, blob)
+	}
+
+	return blobs, nil
+}
+
+func (codec *IndexFileBinlogCodec) DeserializeImpl(blobs []*Blob) (
+	indexBuildID UniqueID,
+	version int64,
+	collectionID UniqueID,
+	partitionID UniqueID,
+	segmentID UniqueID,
+	fieldID UniqueID,
+	indexParams map[string]string,
+	indexName string,
+	indexID UniqueID,
+	datas []*Blob,
+	err error,
+) {
+	if len(blobs) == 0 {
+		return 0, 0, 0, 0, 0, 0, nil, "", 0, nil, errors.New("blobs is empty")
+	}
+	indexParams = make(map[string]string)
+	datas = make([]*Blob, 0)
+
+	for _, blob := range blobs {
+		binlogReader, err := NewBinlogReader(blob.Value)
+		if err != nil {
+			log.Warn("failed to read binlog",
+				zap.Error(err))
+			return 0, 0, 0, 0, 0, 0, nil, "", 0, nil, err
+		}
+		dataType := binlogReader.PayloadDataType
+
+		//desc, err := binlogReader.readDescriptorEvent()
+		//if err != nil {
+		//	log.Warn("failed to read descriptor event",
+		//		zap.Error(err))
+		//	return 0, 0, 0, 0, 0, 0, nil, "", 0, nil, err
+		//}
+		desc := binlogReader.descriptorEvent
+		extraBytes := desc.ExtraBytes
+		extra := make(map[string]interface{})
+		_ = json.Unmarshal(extraBytes, &extra)
+
+		value, _ := strconv.Atoi(extra["indexBuildID"].(string))
+		indexBuildID = UniqueID(value)
+
+		value, _ = strconv.Atoi(extra["version"].(string))
+		version = int64(value)
+
+		collectionID = desc.CollectionID
+		partitionID = desc.PartitionID
+		segmentID = desc.SegmentID
+		fieldID = desc.FieldID
+
+		indexName = extra["indexName"].(string)
+
+		value, _ = strconv.Atoi(extra["indexID"].(string))
+		indexID = UniqueID(value)
+
+		key := extra["key"].(string)
+
+		for {
+			eventReader, err := binlogReader.NextEventReader()
+			if err != nil {
+				log.Warn("failed to get next event reader",
+					zap.Error(err))
+				binlogReader.Close()
+				return 0, 0, 0, 0, 0, 0, nil, "", 0, nil, err
+			}
+			if eventReader == nil {
+				break
+			}
+			switch dataType {
+			// just for backward compatibility
+			case schemapb.DataType_Int8:
+				content, err := eventReader.GetByteFromPayload()
+				if err != nil {
+					log.Warn("failed to get byte from payload",
+						zap.Error(err))
+					eventReader.Close()
+					binlogReader.Close()
+					return 0, 0, 0, 0, 0, 0, nil, "", 0, nil, err
+				}
+
+				if key == IndexParamsKey {
+					_ = json.Unmarshal(content, &indexParams)
+				} else {
+					blob := &Blob{Key: key}
+					// for go reader, no need to copy again
```

Comment:
What if `eventReader` is not of go reader? The quickest way is to make a type assertion here, or the `EventReader` should have a method to tell callers whether to copy data.